### PR TITLE
ci: run only the necessary pre-checks

### PR DESCRIPTION
**Pull request by P403n1x87**
We dynamically generate the list of pre-checks to run on a PR based on the paths included in the PR.

## Checklist

- [x] Change(s) are motivated and described in the PR description.
- [x] Testing strategy is described if automated tests are not included in the PR.
- [x] Risk is outlined (performance impact, potential for breakage, maintainability, etc).
- [x] Change is maintainable (easy to change, telemetry, documentation).
- [x] [Library release note guidelines](https://ddtrace.readthedocs.io/en/stable/releasenotes.html) are followed. If no release note is required, add label `changelog/no-changelog`.
- [x] Documentation is included (in-code, generated user docs, [public corp docs](https://github.com/DataDog/documentation/)).
- [x] Backport labels are set (if [applicable](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting))

## Reviewer Checklist

- [x] Title is accurate.
- [x] No unnecessary changes are introduced.
- [x] Description motivates each change.
- [x] Avoids breaking [API](https://ddtrace.readthedocs.io/en/stable/versioning.html#interfaces) changes unless absolutely necessary.
- [x] Testing strategy adequately addresses listed risk(s).
- [x] Change is maintainable (easy to change, telemetry, documentation).
- [x] Release note makes sense to a user of the library.
- [x] Reviewer has explicitly acknowledged and discussed the performance implications of this PR as reported in the benchmarks PR comment.
- [x] Backport labels are set in a manner that is consistent with the [release branch maintenance policy](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting)

### CODE REVIEW

File: scripts/needs_testrun.py [hunk 1]
```diff
@@ -151,6 +151,11 @@ def for_each_testrun_needed(suites: t.List[str], action: t.Callable[[str], None]
             action(suite)
 
 
+def pr_matches_patterns(patterns: t.Set[str]) -> bool:
```
Comment (Yun-Kim) [inline scripts/needs_testrun.py:154]: Just a question - how do we determine which prechecks are necessary?

File: scripts/gen_circleci_config.py [hunk 2]
```diff
@@ -21,6 +23,61 @@ def gen_required_suites(template: dict) -> None:
...
+    from needs_testrun import pr_matches_patterns
+
+    def check(name: str, command: str, paths: t.Set[str]) -> None:
+        if pr_matches_patterns(paths):
+            template["jobs"]["pre_check"]["steps"].append({"run": {"name": name, "command": command}})
+
+    check(
+        name="Formatting check",
+        command="riot -v run -s fmt && git diff --exit-code",
+        paths={"*.py", "*.pyi"},
```
Comment (P403n1x87) [inline scripts/gen_circleci_config.py:37]: @Yun-Kim an example of how we decide if we should run a check: we provide a set of file patterns. If any of the files within a PR match a pattern then we run the check

Comment (Yun-Kim) [reply scripts/gen_circleci_config.py:37]: ahhhh I missed this. Thanks!

File: scripts/needs_testrun.py (hunk 1 above)
Comment (P403n1x87) [reply scripts/needs_testrun.py:154]: See https://github.com/DataDog/dd-trace-py/pull/6516#discussion_r1276883261

**Review by Yun-Kim — APPROVED** [review]

**Pull request merged**